Use EnhancedStackTrace for spans; start the stack trace on start for EF Core

kanadaj opened the pull request:
Addresses https://github.com/elastic/apm-agent-dotnet/issues/944 : Entity Framework Core stacktraces are now captured at the beginning of the request to avoid async stacktrace shenanigans; I've also replaced the StackTrace from `Span` with `EnhancedStackTrace` to reduce noise in traces.

This might have a performance impact due to capturing stack traces before we can evalute  the span length, but I'm not quite sure if it's possible to get around that. It might be prudent to add a flag to disable trace capture for EF Core. @gregkalapos thoughts?

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-948/detail/PR-948/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-948/detail/PR-948/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-948/detail/PR-948/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp%2FPR-948/detail/PR-948/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #948 opened]`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2020-09-01T15:50:24.018+0000

* Duration: 3 min 54 sec


* Commit: bb5525bf13f546a14fd10e6e8cc963ff00ce987c




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Error signal`
    * Description: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-09-01T15:53:13.393+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/apm-agent-dotnet-mbp/pipelines/PR-948/runs/1/steps/83/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-09-01T15:52:52.440Z] Resetting working tree
[2020-09-01T15:52:52.440Z]  > git reset --hard # timeout=10
[2020-09-01T15:52:52.558Z]  > git clean -fdx # timeout=10
[2020-09-01T15:52:53.893Z] [INFO] Number of builds to be searched 10
[2020-09-01T15:52:54.151Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2020-09-01T15:52:54.181Z] Running in /var/lib/jenkins/workspace/tnet_apm-agent-dotnet-mbp_PR-948/apm-agent-dotnet
[2020-09-01T15:52:54.227Z] [INFO] gitCheckout: Checkout SCM PR-948 with default customisation from the Item.
[2020-09-01T15:52:54.265Z] [INFO] Override default checkout
[2020-09-01T15:52:54.341Z] Sleeping for 10 sec
[2020-09-01T15:53:04.505Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-09-01T15:53:04.527Z] Wiping out workspace first.
[2020-09-01T15:53:04.537Z] Cloning the remote Git repository
[2020-09-01T15:53:04.537Z] Using shallow clone with depth 3
[2020-09-01T15:53:04.537Z] Avoid fetching tags
[2020-09-01T15:53:04.551Z] Cloning repository git@github.com:elastic/apm-agent-dotnet.git
[2020-09-01T15:53:04.575Z]  > git init /var/lib/jenkins/workspace/tnet_apm-agent-dotnet-mbp_PR-948/apm-agent-dotnet # timeout=10
[2020-09-01T15:53:04.588Z] Fetching upstream changes from git@github.com:elastic/apm-agent-dotnet.git
[2020-09-01T15:53:04.588Z]  > git --version # timeout=10
[2020-09-01T15:53:04.599Z]  > git --version # 'git version 2.17.1'
[2020-09-01T15:53:04.599Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-09-01T15:53:04.604Z]  > git fetch --no-tags --progress -- git@github.com:elastic/apm-agent-dotnet.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2020-09-01T15:53:05.772Z] Cleaning workspace
[2020-09-01T15:53:05.788Z] Using shallow fetch with depth 3
[2020-09-01T15:53:05.788Z] Pruning obsolete local branches
[2020-09-01T15:53:05.757Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-dotnet.git # timeout=10
[2020-09-01T15:53:05.760Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-09-01T15:53:05.768Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-dotnet.git # timeout=10
[2020-09-01T15:53:05.774Z]  > git rev-parse --verify HEAD # timeout=10
[2020-09-01T15:53:05.778Z] No valid HEAD. Skipping the resetting
[2020-09-01T15:53:05.778Z]  > git clean -fdx # timeout=10
[2020-09-01T15:53:05.790Z] Fetching upstream changes from git@github.com:elastic/apm-agent-dotnet.git
[2020-09-01T15:53:05.791Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-09-01T15:53:05.794Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-agent-dotnet.git +refs/pull/948/head:refs/remotes/origin/PR-948 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-09-01T15:53:06.567Z] Merging remotes/origin/master commit ffe234e5301fe8d6c71e70e06894fce03baf1c65 into PR head commit bb5525bf13f546a14fd10e6e8cc963ff00ce987c
[2020-09-01T15:53:06.694Z] Merge succeeded, producing bb5525bf13f546a14fd10e6e8cc963ff00ce987c
[2020-09-01T15:53:06.695Z] Checking out Revision bb5525bf13f546a14fd10e6e8cc963ff00ce987c (PR-948)
[2020-09-01T15:53:06.818Z] Commit message: "Use EnhancedStackTrace for spans; start the stack trace on start for EF core"
[2020-09-01T15:53:06.818Z] Cleaning workspace
[2020-09-01T15:53:06.954Z]  > git --version # timeout=10
[2020-09-01T15:53:06.958Z]  > git --version # 'git version 2.17.1'
[2020-09-01T15:53:06.966Z] fatal: bad object 06c62f097a3109e82d977b3292df1e77ce92ff26
[2020-09-01T15:53:06.570Z]  > git config core.sparsecheckout # timeout=10
[2020-09-01T15:53:06.574Z]  > git checkout -f bb5525bf13f546a14fd10e6e8cc963ff00ce987c # timeout=15
[2020-09-01T15:53:06.667Z]  > git remote # timeout=10
[2020-09-01T15:53:06.670Z]  > git config --get remote.origin.url # timeout=10
[2020-09-01T15:53:06.675Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-09-01T15:53:06.678Z]  > git merge ffe234e5301fe8d6c71e70e06894fce03baf1c65 # timeout=10
[2020-09-01T15:53:06.690Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-09-01T15:53:06.697Z]  > git config core.sparsecheckout # timeout=10
[2020-09-01T15:53:06.700Z]  > git checkout -f bb5525bf13f546a14fd10e6e8cc963ff00ce987c # timeout=15
[2020-09-01T15:53:06.821Z]  > git rev-parse --verify HEAD # timeout=10
[2020-09-01T15:53:06.826Z] Resetting working tree
[2020-09-01T15:53:06.826Z]  > git reset --hard # timeout=10
[2020-09-01T15:53:06.930Z]  > git clean -fdx # timeout=10
[2020-09-01T15:53:07.958Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-09-01T15:53:08.646Z] + git fetch https://****:****@github.com/elastic/apm-agent-dotnet.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-09-01T15:53:10.076Z] Archiving artifacts
[2020-09-01T15:53:10.766Z] + git rev-parse HEAD
[2020-09-01T15:53:11.141Z] + git rev-parse HEAD
[2020-09-01T15:53:11.452Z] + git rev-parse origin/pr/948
[2020-09-01T15:53:11.517Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-09-01T15:53:11.883Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-09-01T15:53:12.923Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/apm-agent-dotnet/pulls/948/reviews return 0 elements
[2020-09-01T15:53:13.000Z] [INFO] githubPrCheckApproved: Title: Use EnhancedStackTrace for spans; start the stack trace on start for EF Core - User: kanadaj - Author Association: CONTRIBUTOR
[2020-09-01T15:53:13.414Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2020-09-01T15:53:13.414Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2020-09-01T15:53:13.514Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2020-09-01T15:53:13.547Z] Sleeping for 5 sec
[2020-09-01T15:53:14.759Z] Stage "Parallel" skipped due to earlier failure(s)
[2020-09-01T15:53:14.970Z] Stage "Linux" skipped due to earlier failure(s)
[2020-09-01T15:53:14.976Z] Stage "Windows .NET Framework" skipped due to earlier failure(s)
[2020-09-01T15:53:14.982Z] Stage "Windows .NET Core" skipped due to earlier failure(s)
[2020-09-01T15:53:14.988Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-09-01T15:53:14.999Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2020-09-01T15:53:15.159Z] Failed in branch Integration Tests
[2020-09-01T15:53:15.168Z] Failed in branch Benchmarks
[2020-09-01T15:53:15.169Z] Stage "Linux" skipped due to earlier failure(s)
[2020-09-01T15:53:15.178Z] Stage "Windows .NET Framework" skipped due to earlier failure(s)
[2020-09-01T15:53:15.187Z] Stage "Windows .NET Core" skipped due to earlier failure(s)
[2020-09-01T15:53:15.415Z] Stage "Linux" skipped due to earlier failure(s)
[2020-09-01T15:53:15.424Z] Stage "Windows .NET Framework" skipped due to earlier failure(s)
[2020-09-01T15:53:15.433Z] Stage "Windows .NET Core" skipped due to earlier failure(s)
[2020-09-01T15:53:15.611Z] Failed in branch Linux
[2020-09-01T15:53:15.679Z] Stage "Windows .NET Framework" skipped due to earlier failure(s)
[2020-09-01T15:53:15.686Z] Stage "Windows .NET Core" skipped due to earlier failure(s)
[2020-09-01T15:53:15.833Z] Failed in branch Windows .NET Core
[2020-09-01T15:53:15.884Z] Stage "Windows .NET Framework" skipped due to earlier failure(s)
[2020-09-01T15:53:16.025Z] Failed in branch Windows .NET Framework
[2020-09-01T15:53:16.431Z] Stage "Release to feedz.io" skipped due to earlier failure(s)
[2020-09-01T15:53:16.533Z] Stage "Release" skipped due to earlier failure(s)
[2020-09-01T15:53:16.578Z] Stage "Release" skipped due to earlier failure(s)
[2020-09-01T15:53:16.683Z] Stage "Release" skipped due to earlier failure(s)
[2020-09-01T15:53:16.853Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-09-01T15:53:16.899Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-09-01T15:53:17.273Z] Running on Jenkins in /var/lib/jenkins/workspace/tnet_apm-agent-dotnet-mbp_PR-948
[2020-09-01T15:53:17.446Z] [INFO] getVaultSecret: Getting secrets
[2020-09-01T15:53:17.538Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-09-01T15:53:18.654Z] + chmod 755 generate-build-data.sh
[2020-09-01T15:53:18.655Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-948/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-948/runs/1 ABORTED 174366
[2020-09-01T15:53:19.205Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/apm-agent-dotnet-mbp/PR-948/runs/1/steps/?limit=10000 -o steps-info.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

gregkalapos commented on src/Elastic.Apm.EntityFrameworkCore/EfCoreDiagnosticListener.cs:
```diff
@@ -38,7 +39,8 @@ public void OnNext(KeyValuePair<string, object> kv)
 				case { } k when k == RelationalEventId.CommandExecuting.Name && _agent.Tracer.CurrentTransaction != null:
 					if (kv.Value is CommandEventData commandEventData)
 					{
-						var newSpan = _agent.TracerInternal.DbSpanCommon.StartSpan(_agent, commandEventData.Command, InstrumentationFlag.EfCore);
```
I'd prefer keeping the call to `DbSpanCommon.StartSpan` - that way all db spans are started in the same method. And then I'd just add an optional parameter to it and pass `true` so it calls `StartSpanOnCurrentExecutionSegment` with `captureStackTraceOnStart: true`

kanadaj replied:
Alright, I didn't want to mess with the internal API and HTTP calls use the same method, but I'll change it around then

gregkalapos reviewed (commented):
Great! Thanks for the PR! 
> I've also replaced the StackTrace from Span with EnhancedStackTrace to reduce noise in traces.

Great! This was probably missing from a previous PR, the intention was to use `EnhancedStackTrace` everywhere.

> This might have a performance impact due to capturing stack traces before we can evalute the span length, but I'm not quite sure if it's possible to get around that. It might be prudent to add a flag to disable trace capture for EF Core. @gregkalapos thoughts?

We have the same with HTTP calls and I'd say we can except this. My argument for this is that within a transaction typically you'll have a few such transactions and very likely they'll be longer than the default minduration for capturing stack trace. So it's definitely not like people will have 100s of such spans in a transaction, and if they do, then the perf. impact of this won't be the biggest problem.

So far for HTTP we have no request to disable this behavior, so I'd not add a config yet. If we have request to do so I'd rethink this.

One last point: could you maybe add 1 test with EF Core with some async calls? I think the most important assert would be one asserting that the caller method is on the stack. Let me know if you can't do it - In that case I'll do it.

Other than the test and my comment from before I'd be happy to merge this.

Thanks again!

kanadaj commented:
> One last point: could you maybe add 1 test with EF Core with some async calls? I think the most important assert would be one asserting that the caller method is on the stack. Let me know if you can't do it - In that case I'll do it.

Alright, changed the internal method to add the `captureStackTraceOnStart` parameter and added a test for capturing the stacktrace.

To note, in order to make a test that doesn't test successfully even with the old code, I had to create a table inside the DbContext, since `ExecuteSqlCommandAsync` would bypass the EF Core pipeline almost entirely.

gregkalapos reviewed (approved):
Thank you! - looks ok to me.

I guess the integration test step will fail in the PR which was fixed in https://github.com/elastic/apm-agent-dotnet/pull/949 a few hours ago - no need to rebase just for that. I'll let CI run, if all tests except the integration test step is green, then this ready to merge.
